pkg/sdk/grants_impl.go

```diff
@@ -226,6 +256,138 @@ func (v *grants) Show(ctx context.Context, opts *ShowGrantOptions) ([]Grant, err
+// - revoke operate (it may affect grant ownership call)
+// - grant ownership
+// - if pipe was previously running
+//   - grant operate on pipe to current role
+//   - unpause the pipe with system function
+//   - revoke operate on pipe from current role
+func (v *grants) grantOwnershipOnPipe(ctx context.Context, pipeId SchemaObjectIdentifier, opts *GrantOwnershipOptions) error {
+	// To be able to call ALTER on a pipe to stop its execution,
+	// the current role has to be either the owner of this pipe or be granted with OPERATE privilege.
+	// The code below makes certain checks and takes care of making sure that the current role is privileged enough to act on the pipe and safely grant ownership.
+
+	currentRole, err := v.client.ContextFunctions.CurrentRole(ctx)
+	if err != nil {
+		return err
+	}
+	currentRoleName := NewAccountObjectIdentifier(currentRole)
+
+	currentGrants, err := v.client.Grants.Show(ctx, &ShowGrantOptions{
+		On: &ShowGrantsOn{
+			Object: &Object{
+				ObjectType: ObjectTypePipe,
+				Name:       pipeId,
+			},
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	canOperateOnPipe := slices.ContainsFunc(currentGrants, func(grant Grant) bool {
+		return grant.GranteeName == currentRoleName && (grant.Privilege == "OWNERSHIP" || grant.Privilege == SchemaObjectPrivilegeOperate.String())
+	})
+
+	var revokeOperate func() error
+	if !canOperateOnPipe {
+		revokeOperate, err = v.grantOperateOnPipeTemporarily(ctx, pipeId, currentRoleName)
+		if err != nil {
+			return err
+		}
+	}
+
+	originalPipeExecutionState, err := v.client.SystemFunctions.PipeStatus(pipeId)
+	if err != nil {
+		return err
+	}
+
+	if originalPipeExecutionState == RunningPipeExecutionState {
+		if err := v.client.Pipes.Alter(ctx, pipeId, &AlterPipeOptions{
+			Set: &PipeSet{
+				PipeExecutionPaused: Bool(true),
+			},
+		}); err != nil {
+			return err
+		}
+	}
+
+	if !canOperateOnPipe {
+		if err := revokeOperate(); err != nil {
+			return err
+		}
+	}
+
+	if err := validateAndExec(v.client, ctx, opts); err != nil {
+		return err
+	}
+
+	if originalPipeExecutionState == RunningPipeExecutionState {
+		// We cannot resume right away and "normally" through ALTER, because:
+		// 1. Insufficient privileges (the current role has to be granted with at least OPERATE privilege).
+		// 2. Snowflake throws an error whenever pipes changes its owner, and you try to unpause it. The error suggests using system function to forcefully resume the pipe.
+		revokeOperate, err := v.grantOperateOnPipeTemporarily(ctx, pipeId, currentRoleName)
+		if err != nil {
+			return err
+		}
+
+		// TODO: check if options need to be passed
+		if err := v.client.SystemFunctions.PipeForceResume(pipeId, nil); err != nil {
+			return err
+		}
+
+		if err := revokeOperate(); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (v *grants) grantOperateOnPipeTemporarily(ctx context.Context, pipeId SchemaObjectIdentifier, currentRole AccountObjectIdentifier) (func() error, error) {
```

Comment:
But still there's a case where:
1. Pipe (is running) and the current role we're using is e.g. accountadmin
2. We pause the pipe (right now, it doesn't matter how)
3. We grant ownership on pipe to role X
4. After granting ownership to role X we don't have permission on this pipe (even as accountadmin)
5. To resume the pipe we can either try to use `USE ROLE` on the role X (also hacky because it has to be granted to the user; we would have to enforce that or grant it ourselves) or grant OPERATE privilege on the current role for the time of resuming (like it is right now).

I think other options would require some additional work from the user pov. We can also just leave the pipe paused and figure something out from the perspective of pipe resource or document that pipes need manual work after ownership transfer.

Reply:
I think throwing the error there informing the user that manual work is required is the right option:
- it is not hacky
- nothing happens behind the curtains
- we knew and agreed that such cases will have to be described in the docs, so users are aware that for ownership grants manual work is expected

Reply:
Changed to a non-hacky version, with as much automation as possible (it depends on the user and what privileges are granted to the pipe).